### PR TITLE
Navigation component: Add back button click handler

### PR DESCRIPTION
**Pull request by david-szabo97**
Related: https://github.com/WordPress/gutenberg/issues/25247

## Description
Allow specifying custom back button click handler.

## How has this been tested?
- Run in terminal: `yarn storybook:dev`
- Click **Category**
- Click **Custom back click handler**
- Click **Go back to root**

## Screenshots
image

## Types of changes
Enhancement

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by david-szabo97** [comment]
Back button will be shown if either `parentMenu` or `onBackButtonClick` prop is specified on `NavigationMenu`

I wonder which should take priority. Right now if both `parentMenu` and `onClick` are specified in the `NavigationBackButton` then the `parentMenu` takes priority so the button will act as a back button to the parent menu.

It doesn't make sense to have both props specified, but `parentMenu` might act as an information. For example, when I'm creating menus it would be great to see what's the parent menu of the menu I'm creating. But I'd like to have a custom back button click handler.

IMO, `onClick` should take precedence. Or we can utilise the `event.preventDefault()`.

```js
<NavigationMenu
	menu="custom-back-click"
	title="Custom back button click handler"
	onBackButtonClick={ ( e ) => {
		e.preventDefault(); // This will prevent the default onClick handler which navigates to the parentMenu
		setActiveMenu( 'root' );
	} }
	backButtonLabel="Go back to root"
/>
```

**Review by mattwiebe — APPROVED** [review]
Tests well, code looks good

File: packages/components/src/navigation/stories/index.js [hunk 1]
```diff
@@ -147,6 +147,11 @@ function Example() {
 						item="child-5"
 						title="Automatic back"
 					/>
+					<NavigationItem
+						navigateToMenu="custom-back-click"
+						item="child-5"
```
Comment (vindl) [inline packages/components/src/navigation/stories/index.js:152]: This should be `child-6`.

**Review by vindl — COMMENTED** [review]
Overall this looks good but needs rebase after latest Storybook reorganization from @Copons.

**Comment by vindl** [comment]
> I wonder which should take priority. Right now if both parentMenu and onClick are specified in the NavigationBackButton then the parentMenu takes priority so the button will act as a back button to the parent menu.
...
IMO, onClick should take precedence. Or we can utilise the event.preventDefault().

I also think that onClick should take precedence but not completely override it, since we might want to execute some side effects in it, and then continue with back navigation as usual (e.g. let's say someone wants to track stats for back button clicks in their component or similar). On the other hand, if someone wants a complete override of the behavior, they can optionally use `event.preventDefault()` as you've pointed out.

**Comment by david-szabo97** [comment]
* `onBackButtonClick` will be called with the `Event` object. This way we can use `event.preventDefault()` to prevent the default behavior (navigate back to `parentMenu`)
* Rebased and added stories.

File: packages/components/src/navigation/back-button/index.js [hunk 2]
```diff
@@ -29,15 +29,23 @@ function NavigationBackButton(
 
 	const parentMenuTitle = navigationTree.getMenu( parentMenu )?.title;
 
+	const handleOnClick = ( event ) => {
```
Comment (noahtallen) [inline packages/components/src/navigation/back-button/index.js:32]: Does it make sense to add `useCallback` here to maintain a single reference unless a dependency changes?

Comment (david-szabo97) [reply packages/components/src/navigation/back-button/index.js:32]: I didn't profile it, but I think `useCallback` would make this worse. This function is simple and it is only used in one component, so we would only save 1 rerender which is probably cheaper than `useCallback`. 🤔

Comment (noahtallen) [reply packages/components/src/navigation/back-button/index.js:32]: Sounds good!

File: packages/components/src/navigation/stories/more-examples.js [hunk 3]
```diff
@@ -83,6 +98,33 @@ export function MoreExamplesStory() {
...
+					setBackButtonPreventedBadge( backButtonPreventedBadge + 1 );
+				} }
+				backButtonLabel="Increment badge"
+			>
+				<NavigationItem
+					item="custom-back-click-prevented-child-1"
+					title="You can't go back from here"
+					badge={ backButtonPreventedBadge }
+				/>
+			</NavigationMenu>
```
Comment (Copons) [inline packages/components/src/navigation/stories/more-examples.js:127]: Ahaha this example is EVIL!

I like it.

**Review by Copons — APPROVED** [review]

**Pull request merged**